### PR TITLE
Deactivate Patient Inline

### PR DESCRIPTION
**Pull request by ChristianMurphy**
**Taiga User Story**
TG-634

**Taiga Task(s)**
TG-651

**What did you change?**
Allowed Patient to be deactivated.
Ensure that patient is active in all patient related queries.

**How can we test?**
Go to the patient view.
Click "deactivate"
Go back to the trial view.
Patient should not appear.

**Discussion**
This is based on the design idea originally intended when the current schema was setup.
All tables have a `deletedAt` column, that when a row is soft deleted, will be set to the current date time.

This implementation sets that column when deactivating a patient, and ensures that all queries related to patient ignore deactivated patients.

This is different from the approach that @kgary  suggested where patients would be copied to an external table/database.

I already had some code for it ready to go, so I'm putting it out there as an option.
This could also be redone in the approach suggested by @kgary

Thoughts? :thought_balloon:

### CODE REVIEW

File: controller/handler/dashboard.js [hunk 1]
```diff
@@ -26,6 +26,7 @@ function dashboardView (request, reply) {
         	FROM stage AS st
         	JOIN patient AS pa
         	ON pa.stageId = st.id
+            WHERE pa.deletedAt IS NULL
```
Comment (ChristianMurphy) [inline controller/handler/dashboard.js]: cleanup indent

**Comment by ChristianMurphy** [comment]
This still needs unit tests

**Comment by poojaRal** [comment]
This idea looks great for me now with the current schema. It would be better if @kgary checks this and comments on it. Personally I do not recommend schema change now, which means adding a new table and archiving all deleted info in it.

**Comment by kgary** [comment]
I am not a fan of any approach that requires every query to be modified to filter out deactivated. Sooner or later someone will forget. I do prefer copying it out of the way as it avoids this situation and because rarely when we deactivate will we need that day again (reactivation won't happen often, if ever).
Another possibility is to create a view based on a query to the table(s) that filters out the deactivated patients. I am not certain how many queries and tables we are talking about though.

**Comment by ChristianMurphy** [comment]
> I am not a fan of any approach that requires every query to be modified to filter out deactivated. Sooner or later someone will forget.

I agree this could happen, but I see low risk if it does happen, there would be no data loss and fixing it is a single line code change. :neutral_face: 

> I do prefer copying it out of the way as it avoids this situation and because rarely when we deactivate will we need that day again (reactivation won't happen often, if ever).

I am uncomfortable with the idea of a mirror `deleted` database, it would double the amount of migration needed in eventuality that new tables or columns are added.
Also I see forgetting to migrate the `deleted` database as a greater risk than forgetting to filter out inactive pins. (forgetting to migrate = constraint errors on insert = data loss)

Plus we are in the final stabilization phase for this release, do we really want to add a new database/tables at this point in time? ¯\\\_(ツ)\_/¯

> Another possibility is to create a view based on a query to the table(s) that filters out the deactivated patients.

Applying a view is something I am more comfortable with.
Sequelize automatically applies an 'active only' filter to queries targeting just the `patient` table.
And a custom view could be added to check for only active patients in other queries. :smile: 

> I am not certain how many queries and tables we are talking about though.

10 queries

----

In the end I just want to make sure most of the possible negative side affects are considered.
If a separate database/table still looks like the best option, I'll change the approach :+1:

**Comment by kgary** [comment]
Well, I wasn't talking about a new database, I was talking about a table.
sequelize is not relevant, proper database design is.
Keep in mind that the portal will be migrated to talk to a REST API, not directly to the database in the near future.
Let's do the view solution

**Comment by ChristianMurphy** [comment]
will do :+1:

File: model/view-active-patients.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+ * @returns {Null} nothing
+ */
+function register (sequelize) {
+    /**
+     * patient that has not been deactivated
+     * @typedef {Object} ActivePatient
+     */
+    sequelize.query(
+        `
+        CREATE OR REPLACE VIEW active_patients AS
```
Comment (ChristianMurphy) [inline model/view-active-patients.js]: @promisasu/admin is `active_patients` a good name here?
Or is this going to be confused with a patient who is currently within their start and end date?

**Comment by kgary** [comment]
I made this clear at a previous meeting - I do not want sequelize or any other external ORM mapping thingy in charge of the schema. I do not want any DDL in anything other than an SQL script. Please remove all DDL from sequalize

**Comment by ChristianMurphy** [comment]
The sequelize based DDL is already optional.
the schema can also be generated off this schema script.
https://gist.github.com/ChristianMurphy/c267bb92ffbce982571fc86574686042

The sequelize DDL has been left in thus far because:

1. it is a convenient way to setup the schema
2. sequelize offers helper methods when you also have a json description of the db (even if it is not used to generate the database)
3. its already there

**Comment by kgary** [comment]
I don't know what part of "please remove it" you don't understand  :<P

File: model/view-active-patients.js (hunk 2 above)
Comment (poojaRal) [inline model/view-active-patients.js]: This looks like a good name to me. Active patients are the ones which are currently active in the system

**Comment by poojaRal** [comment]
LGTM

**Comment by iamaureen** [comment]
LGTM

**Pull request merged**